cleanrl/ppo_dna_atari_envpool.py

```diff
@@ -0,0 +1,468 @@
+        probs = Categorical(logits=logits)
+        return probs.log_prob(action), probs.entropy()
+
+    @staticmethod
+    def action_kl_divergence(old_logits, new_logits):
+        old_probs = Categorical(logits=old_logits)
+        new_probs = Categorical(logits=new_logits)
+        return kl_divergence(old_probs, new_probs).mean()
+
+
+def main():
+    args = parse_args()
+    run_name = f"{args.env_id}__{args.exp_name}__{args.seed}__{int(time.time())}"
+    if args.track:
+        import wandb
+
+        wandb.init(
+            project=args.wandb_project_name,
+            entity=args.wandb_entity,
+            sync_tensorboard=True,
+            config=vars(args),
+            name=run_name,
+            monitor_gym=True,
+            save_code=True,
+        )
+    writer = SummaryWriter(f"runs/{run_name}")
+    writer.add_text(
+        "hyperparameters",
+        "|param|value|\n|-|-|\n%s" % ("\n".join([f"|{key}|{value}|" for key, value in vars(args).items()])),
+    )
+
+    # Fix multiplication precision on Ampere devices (30xx) prior to PyTorch 1.12
+    # https://pytorch.org/docs/stable/notes/cuda.html#tf32-on-ampere
+    torch.backends.cuda.matmul.allow_tf32 = False
+
+    # TRY NOT TO MODIFY: seeding
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed)
+    torch.backends.cudnn.deterministic = args.torch_deterministic
+
+    device = torch.device("cuda" if torch.cuda.is_available() and args.cuda else "cpu")
+
+    # env setup
+    envs = envpool.make(
+        args.env_id,
+        env_type="gym",
+        num_envs=args.num_envs,
+        # batch_size=args.epoch_size,  # async
+        num_threads=args.envpool_num_threads,
+        episodic_life=True,
+        seed=args.seed,
+    )
+    envs.is_vector_env = True
+    envs.num_envs = args.num_envs
+    envs.single_action_space = envs.action_space
+    envs.single_observation_space = envs.observation_space
+    envs = RecordEpisodeStatistics(envs)
+    # if args.capture_video:
+    #     envs = gym.wrappers.RecordVideo(envs, f"videos/{run_name}")
+    envs = gym.wrappers.NormalizeObservation(envs)
+    envs = gym.wrappers.TransformObservation(envs, lambda obs: np.clip(obs, -3, 3))
+    envs = gym.wrappers.NormalizeReward(envs, gamma=args.gamma)
+    envs = gym.wrappers.TransformReward(envs, lambda reward: np.clip(reward, -5, 5))
+    assert isinstance(envs.action_space, gym.spaces.Discrete), "only discrete action space is supported"
+
+    agent_policy = Agent(envs).to(device)
+    agent_value = Agent(envs).to(device)
+    policy_optimizer = optim.Adam(agent_policy.parameters(), lr=args.policy_learning_rate, eps=1e-5)
+    value_optimizer = optim.Adam(agent_value.parameters(), lr=args.value_learning_rate, eps=1e-5)
+    distill_optimizer = optim.Adam(agent_policy.parameters(), lr=args.distill_learning_rate, eps=1e-5)
+
+    if args.track:
+        # log gradients
+        wandb.watch(agent_policy)
+        wandb.watch(agent_value)
```

Comment:
Is gradient logging useful for debugging?

Reply:
yeah it's sometimes been useful, esp. when trying to choose the gradient clipping value, but I can remove this to be in line with the baseline PPO style

Reply:
Interesting - could you elaborate on the process of choosing the gradient clipping value?

Reply:
well I just eyeball the gradient norms over the course of training and compare against returns at the same time, and try to guess if SGD updates might be overshooting